File: PhysicsEngine.swift
//
//  PhysicsEngine.swift
//  HockeyLifeSim
//
//  Created by Brian Google on 8/7/25.
//

import Foundation
import CoreGraphics

class PhysicsEngine {

    // The main entry point for the physics simulation.
    // It takes the current state and a time step (e.g., 1/60th of a second) and returns the new state.
    func update(gameState: EngineGameState, deltaTime: TimeInterval) -> EngineGameState {
        var newState = gameState
        let allPlayers = newState.homeSkaters + newState.awaySkaters + [newState.homeGoalie, newState.awayGoalie]

        // 1. Apply environmental forces like friction.
        applyForces(to: &newState)
        
        // 2. Update the position of every entity based on its velocity.
        moveEntities(in: &newState, deltaTime: deltaTime)
        
        // 3. Detect and resolve any collisions that occurred after moving.
        handleCollisions(in: &newState, allPlayers: allPlayers)
        
        return newState
    }
    
    // MARK: - Core Physics Steps

    private func applyForces(to gameState: inout EngineGameState) {
        let friction = PhysicsCoefficients.iceFriction
        
        // Apply ice friction to the puck
        gameState.puck.velocity.dx *= friction
        gameState.puck.velocity.dy *= friction
        
        // Apply ice friction to all players
        for i in 0..<gameState.homeSkaters.count {
            gameState.homeSkaters[i].velocity.dx *= friction
            gameState.homeSkaters[i].velocity.dy *= friction
        }
        for i in 0..<gameState.awaySkaters.count {
            gameState.awaySkaters[i].velocity.dx *= friction
            gameState.awaySkaters[i].velocity.dy *= friction
        }
        gameState.homeGoalie.velocity.dx *= friction
        gameState.homeGoalie.velocity.dy *= friction
        gameState.awayGoalie.velocity.dx *= friction
        gameState.awayGoalie.velocity.dy *= friction
    }
    
    private func moveEntities(in gameState: inout EngineGameState, deltaTime: TimeInterval) {
        let dt = CGFloat(deltaTime)
        
        // Update puck position
        gameState.puck.position.x += gameState.puck.velocity.dx * dt
        gameState.puck.position.y += gameState.puck.velocity.dy * dt
        
        // Update player positions
        for i in 0..<gameState.homeSkaters.count {
            gameState.homeSkaters[i].position.x += gameState.homeSkaters[i].velocity.dx * dt
            gameState.homeSkaters[i].position.y += gameState.homeSkaters[i].velocity.dy * dt
        }
        for i in 0..<gameState.awaySkaters.count {
            gameState.awaySkaters[i].position.x += gameState.awaySkaters[i].velocity.dx * dt
            gameState.awaySkaters[i].position.y += gameState.awaySkaters[i].velocity.dy * dt
        }
        gameState.homeGoalie.position.x += gameState.homeGoalie.velocity.dx * dt
        gameState.homeGoalie.position.y += gameState.homeGoalie.velocity.dy * dt
        gameState.awayGoalie.position.x += gameState.awayGoalie.velocity.dx * dt
        gameState.awayGoalie.position.y += gameState.awayGoalie.velocity.dy * dt
    }
    
    private func handleCollisions(in gameState: inout EngineGameState, allPlayers: [PlayerBody]) {
        // --- Puck Collisions ---
        handlePuckWallCollision(puck: &gameState.puck)
        handlePuckGoalCollision(puck: &gameState.puck, goalFrame: gameState.homeGoalFrame)
        handlePuckGoalCollision(puck: &gameState.puck, goalFrame: gameState.awayGoalFrame)
        
        // --- Entity-to-Entity Collisions (Player-Player, Puck-Player) ---
        // This is a naive O(n^2) check. For 12 players and 1 puck, this is perfectly fine.
        var bodies: [PhysicsBody] = allPlayers.map { $0 } + [gameState.puck]
        
        for i in 0..<bodies.count {
            for j in (i + 1)..<bodies.count {
                var body1 = bodies[i]
                var body2 = bodies[j]
                
                let dx = body2.position.x - body1.position.x
                let dy = body2.position.y - body1.position.y
                let distance = sqrt(dx*dx + dy*dy)
                
                if distance < body1.radius + body2.radius {
                    // Collision detected! Resolve it.
                    resolveCollision(body1: &body1, body2: &body2)
                    
                    // IMPORTANT: Write the changed data back to the array.
                    bodies[i] = body1
                    bodies[j] = body2
                }
            }
        }
        
        // After iterating, update the main game state from the modified `bodies` array.
        // This is crucial because structs are value types.
        gameState.puck = bodies.first(where: { $0.id == gameState.puck.id }) as! PuckBody
        // (Similar update logic for players would go here)
    }

    // MARK: - Collision Resolution
    
    private func resolveCollision(body1: inout PhysicsBody, body2: inout PhysicsBody) {
        // This function calculates the physically accurate response for a 2D elastic collision.
        let dx = body2.position.x - body1.position.x
        let dy = body2.position.y - body1.position.y
        let distance = sqrt(dx*dx + dy*dy)
        
        // Collision normal vector (points from body1 to body2)
        let normalX = dx / distance
        let normalY = dy / distance
        
        // Relative velocity
        let relativeVelocityX = body1.velocity.dx - body2.velocity.dx
        let relativeVelocityY = body1.velocity.dy - body2.velocity.dy
        
        // Dot product of relative velocity and the normal vector.
        let speed = relativeVelocityX * normalX + relativeVelocityY * normalY
        
        // Only resolve if objects are moving towards each other.
        if speed < 0 { return }

        // The impulse (change in momentum) calculation.
        // Simplified for perfect elasticity for now.
        let impulse = 2 * speed / (body1.mass + body2.mass)
        
        // Apply the impulse to each body's velocity.
        body1.velocity.dx -= impulse * body2.mass * normalX
        body1.velocity.dy -= impulse * body2.mass * normalY
        
        body2.velocity.dx += impulse * body1.mass * normalX
        body2.velocity.dy += impulse * body1.mass * normalY
    }

    private func handlePuckWallCollision(puck: inout PuckBody) {
        // A more advanced implementation that handles rounded corners will be added later.
        if (puck.position.y - puck.radius < 0 && puck.velocity.dy < 0) || (puck.position.y + puck.radius > RinkConstants.height && puck.velocity.dy > 0) {
            puck.velocity.dy *= -PhysicsCoefficients.puckRestitution
        }
        if (puck.position.x - puck.radius < 0 && puck.velocity.dx < 0) || (puck.position.x + puck.radius > RinkConstants.width && puck.velocity.dx > 0) {
            puck.velocity.dx *= -PhysicsCoefficients.puckRestitution
        }
    }
    
    private func handlePuckGoalCollision(puck: inout PuckBody, goalFrame: CGRect) {
        // Simplified: check for intersection with the goal's rectangle.
        // A full implementation would check against the 3 line segments (posts and crossbar).
        let puckFrame = CGRect(x: puck.position.x - puck.radius, y: puck.position.y - puck.radius, width: puck.radius*2, height: puck.radius*2)
        
        if puckFrame.intersects(goalFrame) {
            // High-restitution "ping" off the post.
            // For now, we'll just reverse velocity as a placeholder.
            puck.velocity.dx *= -1.5 // Extra bounce
            puck.velocity.dy *= -1.5
        }
    }
}

File: physics-engine.js
// Physics Engine for Hockey Game Simulation
class Vector2 {
    constructor(x = 0, y = 0) {
        this.x = x;
        this.y = y;
    }

    add(v) {
        return new Vector2(this.x + v.x, this.y + v.y);
    }

    subtract(v) {
        return new Vector2(this.x - v.x, this.y - v.y);
    }

    multiply(scalar) {
        return new Vector2(this.x * scalar, this.y * scalar);
    }

    length() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    normalize() {
        const len = this.length();
        if (len === 0) return new Vector2(0, 0);
        return new Vector2(this.x / len, this.y / len);
    }

    dot(v) {
        return this.x * v.x + this.y * v.y;
    }
}

class PhysicsBody {
    constructor(x, y, radius, mass = 1) {
        this.position = new Vector2(x, y);
        this.velocity = new Vector2(0, 0);
        this.radius = radius;
        this.mass = mass;
        this.friction = 0.98;
    }

    applyForce(force) {
        const acceleration = force.multiply(1 / this.mass);
        this.velocity = this.velocity.add(acceleration);
    }

    update(deltaTime) {
        this.position = this.position.add(this.velocity.multiply(deltaTime));
        this.velocity = this.velocity.multiply(this.friction);
    }
}

class Player extends PhysicsBody {
    constructor(x, y, team, role) {
        super(x, y, 12, 85); // radius 12px, mass 85kg
        this.team = team; // 'home' or 'away'
        this.role = role; // 'C', 'LW', 'RW', 'LD', 'RD', 'G'
        this.hasPuck = false;
        this.speed = 150; // pixels per second
    }

    skateTowards(target, deltaTime) {
        const direction = target.subtract(this.position).normalize();
        const force = direction.multiply(this.speed);
        this.velocity = force.multiply(deltaTime);
    }
}

class Puck extends PhysicsBody {
    constructor(x, y) {
        super(x, y, 5, 0.17); // radius 5px, mass 0.17kg
        this.owner = null;
        this.friction = 0.96;
    }
}

class PhysicsEngine {
    constructor(width, height) {
        this.width = width;
        this.height = height;
        this.rinkBounds = {
            minX: 25,
            maxX: width - 25,
            minY: 25,
            maxY: height - 25
        };
    }

    update(players, puck, goalies, deltaTime) {
        // Update all entities
        players.forEach(player => player.update(deltaTime));
        goalies.forEach(goalie => goalie.update(deltaTime));
        puck.update(deltaTime);

        // Keep entities within bounds
        this.constrainToBounds(players, goalies, puck);

        // Check collisions
        this.checkCollisions(players, goalies, puck);
    }

    constrainToBounds(players, goalies, puck) {
        const allEntities = [...players, ...goalies, puck];
        allEntities.forEach(entity => {
            if (entity.position.x - entity.radius < this.rinkBounds.minX) {
                entity.position.x = this.rinkBounds.minX + entity.radius;
                entity.velocity.x *= -0.5;
            }
            if (entity.position.x + entity.radius > this.rinkBounds.maxX) {
                entity.position.x = this.rinkBounds.maxX - entity.radius;
                entity.velocity.x *= -0.5;
            }
            if (entity.position.y - entity.radius < this.rinkBounds.minY) {
                entity.position.y = this.rinkBounds.minY + entity.radius;
                entity.velocity.y *= -0.5;
            }
            if (entity.position.y + entity.radius > this.rinkBounds.maxY) {
                entity.position.y = this.rinkBounds.maxY - entity.radius;
                entity.velocity.y *= -0.5;
            }
        });
    }

    checkCollisions(players, goalies, puck) {
        const allPlayers = [...players, ...goalies];

        // Check puck-player collisions
        allPlayers.forEach(player => {
            const distance = player.position.subtract(puck.position).length();
            if (distance < player.radius + puck.radius + 10) {
                if (!puck.owner) {
                    puck.owner = player;
                    player.hasPuck = true;
                }
            }
        });

        // Check player-player collisions
        for (let i = 0; i < allPlayers.length; i++) {
            for (let j = i + 1; j < allPlayers.length; j++) {
                this.resolveCollision(allPlayers[i], allPlayers[j]);
            }
        }
    }

    resolveCollision(bodyA, bodyB) {
        const delta = bodyB.position.subtract(bodyA.position);
        const distance = delta.length();
        const minDistance = bodyA.radius + bodyB.radius;

        if (distance < minDistance && distance > 0) {
            const normal = delta.normalize();
            const overlap = minDistance - distance;
            
            // Separate bodies
            const separation = normal.multiply(overlap / 2);
            bodyA.position = bodyA.position.subtract(separation);
            bodyB.position = bodyB.position.add(separation);

            // Elastic collision response
            const relativeVelocity = bodyB.velocity.subtract(bodyA.velocity);
            const velocityAlongNormal = relativeVelocity.dot(normal);

            if (velocityAlongNormal < 0) return;

            const restitution = 0.4;
            const impulse = -(1 + restitution) * velocityAlongNormal;
            const impulseDivided = impulse / (bodyA.mass + bodyB.mass);

            const impulseVector = normal.multiply(impulse);
            bodyA.velocity = bodyA.velocity.subtract(impulseVector.multiply(1 / bodyA.mass));
            bodyB.velocity = bodyB.velocity.add(impulseVector.multiply(1 / bodyB.mass));
        }
    }

    checkGoal(puck, homeGoalArea, awayGoalArea) {
        // Check if puck is in goal
        const puckX = puck.position.x;
        const puckY = puck.position.y;

        if (this.isInGoal(puckX, puckY, homeGoalArea)) {
            return 'away'; // Away team scored
        }
        if (this.isInGoal(puckX, puckY, awayGoalArea)) {
            return 'home'; // Home team scored
        }
        return null;
    }

    isInGoal(x, y, goalArea) {
        return x >= goalArea.x &&
               x <= goalArea.x + goalArea.width &&
               y >= goalArea.y &&
               y <= goalArea.y + goalArea.height;
    }
}
